### PR TITLE
fix: repair training

### PR DESCRIPTION
**Pull request by rjwalters**
image

green is main; red is this PR -- this patch restores loss behavior and performance

## Summary

This PR refactors the MettaAgent's forward pass to separate inference and training modes into distinct methods and calls attention to the different tensor shapes expected. It also splits the monolithic `sample_logits` function from distribution_utils into `sample_actions` and `evaluate_actions` functions.

### 1. MettaAgent Forward Pass Refactoring

**Before**: Single `forward()` method handled both inference and training with complex conditional logic
**After**: Clean separation into specialized methods:

- `forward()` - Main entry point that delegates to appropriate specialized method
- `forward_inference()` - Handles inference mode (action=None) 
- `forward_training()` - Handles training/BPTT mode (action provided)

### 2. Distribution Utils Refactoring

**Before**: Single `sample_logits()` function with optional parameter
```python
def sample_logits(logits, bptt_logit_index=None) -> Tuple[...]
```

**After**: Two specialized functions with clear purposes
```python
def sample_actions(action_logits) -> Tuple[actions, log_probs, entropy, action_log_probs]
```
```python
def evaluate_actions(action_logits, actions) -> Tuple[log_probs, entropy, action_log_probs] 
```

### Forward Pass Flow

**Inference Mode** (`action=None`):
1. Input: `x` shape `(BT, *obs_shape)`
2. LSTM processing produces `value` and `logits` both shape `(BT, *)`
3. `sample_actions()` samples new actions from policy
4. Return sampled actions and probabilities

**Training Mode** (`action` provided):
1. Input: `x` shape `(B, T, *obs_shape)`, `action` shape `(B, T, 2)`
2. LSTM internally reshapes to `(BT, *)` for processing
3. `evaluate_actions()` evaluates provided actions against current policy
4. Return original actions with updated probabilities

### CODE REVIEW

File: metta/agent/metta_agent.py [hunk 1]
```diff
@@ -1,7 +1,6 @@
 import logging
 from typing import List, Optional, Union
 
-import einops
```
Comment (rjwalters) [inline metta/agent/metta_agent.py:4]: I would like to add einops in a followup PR to isolate any possible performance changes

File: metta/agent/util/distribution_utils.py [hunk 2]
```diff
@@ -7,43 +7,73 @@
...
-    # Entropy: -∑p * log(p)
-    entropy = -torch.sum(probs * log_probs, dim=-1)  # Shape: [B*T]
-
-    return action_index, action_log_prob, entropy, log_probs
+
+    action_log_probs = F.log_softmax(action_logits, dim=-1)  # [batch_size, num_actions]
+    action_probs = torch.exp(action_log_probs)  # [batch_size, num_actions]
+
+    # Extract log-probabilities for the provided actions using advanced indexing
+    batch_indices = torch.arange(actions.shape[0], device=actions.device)
```
Comment (Al-does) [inline metta/agent/util/distribution_utils.py:73]: I _thought_ torch will automatically send this new tensor to the same device that actions is on so the device call might be redundant.

**Review by Al-does — APPROVED** [review]

File: metta/agent/metta_agent.py [hunk 3]
```diff
@@ -174,18 +173,85 @@ def lstm(self):
...
+        Forward pass for training mode - evaluates the policy on provided actions.
+
+        Args:
+            value: Value estimate tensor, shape (BT, 1)
+            logits: Action logits tensor, shape (BT, A)
+            action: Action tensor for evaluation, shape (B, T, 2)
+
+        Returns:
+            Tuple of (action, action_log_prob, entropy, value, log_probs)
+            - action: Same as input action, shape (B, T, 2)
```
[the remark this answers is not in the document]
Comment (rjwalters) [reply metta/agent/metta_agent.py:224]: I agree but we need to update how it is called to drop the unused return value -- I want to get away from policy call signatures anyway

File: metta/agent/metta_agent.py [hunk 4]
```diff
@@ -194,56 +260,31 @@ def forward(self, x: torch.Tensor, state: PolicyState, action: Optional[torch.Te
...
-            if __debug__:
+        if __debug__:
+            # Default values in case obs_shape is not available
+            obs_w, obs_h, features = "W", "H", "F"
+
+            # Check if agent_attributes exists, is not None, and contains obs_shape
+            if (
+                hasattr(self, "agent_attributes")
+                and self.agent_attributes is not None
+                and "obs_shape" in self.agent_attributes
```
[the remark this answers is not in the document]
Comment (rjwalters) [reply metta/agent/metta_agent.py:272]: good catch! thanks

File: metta/agent/util/distribution_utils.py (hunk 2 above)
Comment (rjwalters) [reply metta/agent/util/distribution_utils.py:73]: goal here was to capture a slight performance improvement vs creating and then moving

File: metta/agent/metta_agent.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rjwalters) [reply metta/agent/metta_agent.py:272]: actually it appears that in "agent attributes" is it "obs_shape" vs self._obs_shape... we should try to simplify this eventually
